### PR TITLE
Fix infinite blank cycles

### PR DESCRIPTION
**Pull request by louen**
## Summary
This PR fixes an infinite recursion in hand score and a few other miscellaneous issues in the optimization script

## Infinite recursion
The main proposal of this PR is to fix a nasty bug in scoring the following hand:
- **_Great Flood_**
- **_Blizzard_**
- **_Wildfire_**
- **_Book of Changes_** (Changing _Blizzard_ to e.g Beast)
- **_Forest_**(or any land)

In this case, there is a 3-cycle of blanking cards : 
- _Great Flood_ is blanked by _Blizzard_ (blanks all floods)
- _Blizzard_(modified suit to Beast) is blanked by _Wildfire_ (since it isn't a weather anymore)
- _Wildfire_ is blanked by _Great Flood_ (blank all flames except _Lightning_)

The current code scores the [hand of 4](https://fantasy-realms.github.io/index.html?hand=FR08,FR12,FR16,FR49+FR49:FR12:beast) (without the _Forest_) by blanking all three cards in the cycle simultaneously, leaving only the _Book of Changes_ active. (See issue #37) This is consistent with the "Two Basilisks" rule of cards blanking each other in the rules Q&A. 

However, the script fails if you add another Land to the hand (for example a _Forest_).  If you run [this hand](https://fantasy-realms.github.io/index.html?hand=FR04,FR08,FR12,FR16,FR49+FR49:FR12:beast) on the website it will crash with a "too much recursion" error.

The problem is in the `_cardBlanked()` method which will go in a infinite recursion when trying to determine if the _Forest_ is blanked or not:.
There is a piece of code responsible for breaking the cycle by checking if `by === target`, but in this case it doesn't get triggered if the starting card (`target`, i.e. _Forest_) does not belong to the cycle.

I changed the method to maintain a list of traversed cards, which allows to break out of any cycle.

This has the consequence of making a ruling (not explicitly stated in the rules booklet) : in this case **the _Forest_ is counted as active**, as it cannot be blanked by the Flood (since the 3-cycle is entirely blanked).

## Other fixes
I fixed a few issues  in of `find-scores` and `_cardBlanked`
-  a `try catch` block around the scoring (to avoid blocking the page on an infinite recursion)
- a conversion to string in the `normalizeID` to accept integer arguments

I rewrote `_cardBlanked` during my investigation to understand better how it worked. However the logic should be exactly the same except for the bugfix mentioned above.

## Tests
I added a few unit tests around this new edge case.

### CODE REVIEW

File: js/deck.js [hunk 1]
```diff
@@ -1496,6 +1496,10 @@ var deck = {
     }
   },
   getCardById: function (id) {
+    if (typeof(id) == "number") {
```
Comment (louen) [inline js/deck.js:1499]: Bugfix : accept integers in addition to strings.

File: js/find-scores.js [hunk 2]
```diff
@@ -46,71 +47,77 @@ async function findMaxAndMinScore() {
...
     var baseCombinations = [combination];
-    if (necromancer && combination.includes(NECROMANCER)) {
-      var necromancerTargetCards = deck.getCardsBySuit(deck.getCardById(NECROMANCER).relatedSuits);
-      for (const suit of Object.keys(necromancerTargetCards)) {
-        for (const extraCard of necromancerTargetCards[suit]) {
-          if (!combination.includes(extraCard.id)) {
-            var handWithExtraCard = combination.slice();
-            handWithExtraCard.push(extraCard.id);
-            baseCombinations.push(handWithExtraCard);
+    try {
```
Comment (louen) [inline js/find-scores.js:50]: Add try-catch block to scoring.

File: js/find-scores.js [hunk 3]
```diff
@@ -172,7 +179,7 @@ function generateActionVariations(hand) {
     var islandActions = [];
     for (const card of hand.cards()) {
       if (((card.suit === 'Flood' || card.suit === 'Flame' || isPhoenix(card) || hand.containsId(BOOK_OF_CHANGES)) && card.penalty) || card.id === DOPPELGANGER) {
-        islandActions.push([ISLAND, + card.id]);
+        islandActions.push([ISLAND, card.id]);
```
Comment (louen) [inline js/find-scores.js:182]: Bugfix : Fix island action (was NaN)

File: js/hand.js [hunk 4]
```diff
@@ -48,6 +48,10 @@ class Hand {
   }
 
   _normalizeId(id) {
+    if (typeof(id) == "number") {
```
Comment (louen) [inline js/hand.js:51]: Bugfix : accept integers as id

File: js/hand.js [hunk 5]
```diff
@@ -236,18 +240,63 @@ class Hand {
...
+    // Detect mutual blanking and cycles
+    var effectiveBlankers = []
+    for (const by of blankers) {
+      // Card and by mutually blank eachother
+      if (card.blanks !== undefined && card.blanks(by, this)) {
+        return true;
+      }
+
+      // Card and by are part of a cycle of blanks
+      if (stack.includes(by)) {
```
Comment (louen) [inline js/hand.js:288]: This is the main change to the logic of `cardBlanked` which allows to break from the cycle.

**Comment by jpraet** [comment]
Looks good, thanks.
